### PR TITLE
RLogger selection

### PR DESCRIPTION
**Pull request by Axel-Naumann**
```
void func() {
   RLogVerbosityRAII debugThis("FOO", ELogLevel::kDebug);
   Foo::SomethingBroken();
}
```

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on mac11.0/default because I changed your config and want to test

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on mac11.0/default because I might have fixed your config

**Review by eguiraud — APPROVED** [review]
Niiiice, can't wait to use this :smile:

So I guess the pattern to substitute `if (gDebug > 0) Info(...)` would be

```cpp
// once, e.g. before the event loop, to avoid taking locks in all threads all the time just to
// find out that nothing should be emitted most of the time 
const auto fLogLevel = RLogManager::Get().GetVerbosity("RDF");

// in each thread
if (fLogLevel <= kInfo)
  R__INFO__HERE("RDF") << "doing something in this thread";
```

Does that sound reasonable? Or should we check `RLogManager::Get().GetVerbosity("RDF")` in each TBB task?

More comments inline, all minor.

**Comment by Axel-Naumann** [comment]
Will go through your review comments, but first things first: `if (gDebug > 0) printf(...)` is to be replaced by
```
  R__DEBUG_HERE("RDF") << "doing something in this thread";  
```
The fact that you use R__**DEBUG**_HERE signals to `RLogger` that it should only emit this if the verbosity is >= debug.

**Comment by eguiraud** [comment]
> The fact that you use R__DEBUG_HERE signals to RLogger that it should only emit this if the verbosity is >= debug.

@Axel-Naumann yes but to figure out whether verbosity (for the group) is >= debug, a "raw" `R__DEBUG_HERE` will take some locks and go through quite some logic won't it?

**Comment by Axel-Naumann** [comment]
Sure, but your code does exactly the same, only with more code lines ;-)

**Comment by eguiraud** [comment]
I don't think so? I was proposing to retrieve the verbosity level once, before the RDF event loop, and then only call `R__DEBUG_HERE` if the verbosity level requires it (and the check on the cached verbosity level would not lock, so in the common case you'd have no locking). This is at the cost of ignoring changes in the verbosity level that occur during the event loop, of course. I'm open to better solutions :smile:

But I don't think it's reasonable, as a pattern, to unconditionally take locks every time we _might_ want to log something, it would mean that RDF cannot ever put a `R__DEBUG_HERE` or `R__INFO_HERE` in event loop code. `if (gDebug > 0) Info(...)` does not have that issue.

**Comment by pcanal** [comment]
> if (gDebug > 0) Info(...) does not have that issue.

To be correct, gDebug (and it's replacement) should be an atomic.

Also (I probably missed something) are we replacing the current multi-level gDebug by a simple bool? (I.e. either nothing or a flood of a data)?

**Comment by Axel-Naumann** [comment]
> Also (I probably missed something) are we replacing the current multi-level gDebug by a simple bool? (I.e. either nothing or a flood of a data)?

There are verbosity levels (fatal...debug), both global and per "channel".

**Comment by eguiraud** [comment]
> To be correct, gDebug (and it's replacement) should be an atomic.

Depending on the performance impact of checking an atomic in that many places in ROOT, from potentially hundreds of threads, it might be good that gDebug is just an int...

**Comment by pcanal** [comment]
> There are verbosity levels (fatal...debug), both global and per "channel".

Yes, so that is a decrease in functionality.  We currently use both the verbosity level (fatal..debug) and for debug with also have a sub-verbosity level (technically 0 through max int)

**Comment by Axel-Naumann** [comment]
> so that is a decrease in functionality

Yeah that's something that can be fixed. I'm waiting for your brilliant solution to the issue described by Enrico: how do we deal with thread safety (esp of the channel map) vs speed? A global, thread-racy "nobody wants more verbosity than this"?

**Comment by pcanal** [comment]
It is just debug mode, the consequence of a race condition on gDebug would "only" be that some message may or may not be printed when intended.  So it is likely fine.

**Comment by Axel-Naumann** [comment]
For @osschar @linev @bellenot : what I want to hear from you is whether the change in log channel name is okay for you:
- "webeve" => "Eve"
- "WebDisplay", "webgui", "CanvasPainter" => "WebGUI"
- "gpadv7", "Gpadv7", "drawable", "Gpad", "rstyle",  => "GPad"
- "Browser7", "rbrowser" => "Browser"

**Comment by linev** [comment]
> what I want to hear from you is whether the change in log channel name is okay for you:

Yes, one can change names as you propose. Up to now they was used for error messages, where source of problem can be seen anyway.

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 1]
```diff
@@ -37,7 +37,51 @@ enum class ELogLevel {
...
+   std::string fGroup;
+   std::string fFile;
+   std::string fFuncName;
+   int fLine = 0;
+   ELogLevel fLevel;
+
+public:
+   RLogEntry() = default;
+   RLogEntry(ELogLevel level, std::string_view group): fGroup(group), fLevel(level) {}
+   RLogEntry(ELogLevel level, std::string_view group, std::string_view filename, int line, std::string_view funcname)
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx]: Removed `string_view`.

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 2]
```diff
@@ -37,7 +37,51 @@ enum class ELogLevel {
...
+      : fGroup(group), fFile(filename), fFuncName(funcname), fLine(line), fLevel(level)
+   {}
+
+   RLogEntry &SetFile(const std::string &file)
+   {
+      fFile = file;
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:126]: I decided to not touch it for simplicity...

**Comment by bellenot** [comment]
> what I want to hear from you is whether the change in log channel name is okay for you:

Fine with me

**Comment by Axel-Naumann** [comment]
I believe I have addressed all your comments:
```
R__DEBUG_HERE(gRDFLog) << "This foo has " << TakesAWhileToCompute() << " bars.";
```
does not take locks and does not call `TakesAWhileToCompute()` if the verbosity (either global or `gRDFLog`'s) are less than `kDebug`. The verbosity comparisons are racy but that's what we want.

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 3]
```diff
@@ -82,90 +212,101 @@ public:
...
+    RLogScopedVerbosity silence(RLogLevel::kFatal);
+    R__DEBUG_HERE({}) << WillNotBeCalled();
+    ```
+    */
+   struct TagForDefaultVerbosity {};
+   inline bool ShouldCreateLogEntry(ELogLevel level, TagForDefaultVerbosity)
    {
-      fFile = file;
-      return *this;
+      return RLogManager::Get().GetVerbosity() >= level || level < ELogLevel::kInfo;
```
Comment (pcanal) [inline core/foundation/inc/ROOT/RLogger.hxx:302]: How do you suppress Warning message?

Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:302]: `RLogScopedVerbosity silence(ELogLevel::kError)`, unless the channel has warnings turned on. (That's just the answer to your question - we can discuss the interplay of global vs channel verbosity.)

Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:302]: Can you rephrase? I don't understand the second sentence. I'm okay with the first one ;-)

Comment (pcanal) [reply core/foundation/inc/ROOT/RLogger.hxx:302]: Users might be surprised that `RLogManager::Get().SetVerbosity(ELogLevel::kError)` leads to warning message being still printed while `RLogScopedVerbosity silence(ELogLevel::kError)` put in the `main function` removes them.  (Unless I misunderstood something)

Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:302]: `RLogScopedVerbosity silence(ELogLevel::kError)` is just two calls to `RLogManager::Get().SetVerbosity(ELogLevel::kError)`, taking care of resetting to whatever the verbosity was before. I.e.
```
void f() {
  RLogScopedVerbosity silence(ELogLevel::kError);
  Sometihng();
}
```
and
```
void f() {
  auto prev = RLogManager::Get().SetVerbosity(ELogLevel::kError);
  Something();
  RLogManager::Get().SetVerbosity(prev);
}
```
are strictly identical.

Whether a channel's log entries are emitted or not is an "OR" of the global and the channel-specific verbosity: whatever is more verbose wins. Example:

```
RLogChannel ch("Test");
R__WARNING_HERE(ch) << "Emitted";
{
  RLogScopedVerbosity(kError);
  R__WARNING_HERE(ch) << "Suppressed";
}
{
  RLogScopedVerbosity(ch, kError);
  R__WARNING_HERE(ch) << "Emitted"; // global wins - surprisingly?
}
{
  RLogScopedVerbosity(kError);
  RLogScopedVerbosity(ch, kWarning);
  R__WARNING_HERE(ch) << "Visible"; // this time, channel setting wins
}
```

Rationale: allow people to get a silent ROOT - the global verbosity set to `kFatal` should really only show fatal errors unless verbosity is increased for specific channels in which case these channels' logs will be emitted.

One could also argue that a channel's specific setting wins over the global one, and introduce an "unset" channel verbosity, in which case the global verbosity it taken:

```
RLogChannel ch("Test");
R__WARNING_HERE(ch) << "Emitted";
{
  RLogScopedVerbosity(kError);
  R__WARNING_HERE(ch) << "Suppressed";
}
{
  RLogScopedVerbosity(ch, kError);
  R__WARNING_HERE(ch) << "Suppressed"; // channel setting always wins
}
{
  RLogScopedVerbosity(kError);
  RLogScopedVerbosity(ch, kWarning);
  R__WARNING_HERE(ch) << "Visible";
}
```

My feeling was that it might be easier to tell users: "can you run once with `RLogManager::Get().SetVerbosity(ELogLevel::kDebug);`? What's the output?" instead of having to turn on a couple of channels. But I guess that can be done in either scenario. So - shall I change to the second approach because it's less surprising and still allows to do the same things (@pcanal @eguiraud @jblomer mostly, I guess)?

Comment (eguiraud) [reply core/foundation/inc/ROOT/RLogger.hxx:302]: I don't have a lot of experience with logging but "channel-specific wins over global" seems strictly more useful than "more verbose wins" because it _also_ allows running e.g. with "everything on _except_ IO".

However like Philippe in the first comment I'm also confused by the `|| level < ELogLevel::kInfo;`: why do we unconditionally want to create the `RLogEntry` if it's a warning or an error (even if we later won't print them out)? The line probably deserves a comment.

File: core/foundation/src/RLogger.cxx [hunk 4]
```diff
@@ -14,29 +14,33 @@
...
 public:
    // Returns false if further emission of this log entry should be suppressed.
-   bool Emit(const ROOT::Experimental::RLogEntry &entry) override;
+   bool Emit(const RLogEntry &entry) override;
 };
 
-bool RLogHandlerDefault::Emit(const ROOT::Experimental::RLogEntry &entry)
+bool RLogHandlerDefault::Emit(const RLogEntry &entry)
 {
    constexpr static std::array<const char *, 5> sTag{{"Debug", "Info", "Warning", "Log", "FATAL"}};
```
Comment (eguiraud) [inline core/foundation/src/RLogger.cxx:38]:
```suggestion
   constexpr static std::array<const char *, 5> sTag{{"Debug", "Info", "Warning", "Error", "FATAL"}};
```

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 5]
```diff
@@ -176,13 +317,27 @@ public:
...
+#define R__DEBUG_HERE(CHANNEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kDebug, CHANNEL)
+///\}
+
+///\{
+/// \name Macros to create an RLogEntry with specific severity levels auch as `kError + 5`
+#define R__NFATAL_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kFatal + LEVEL, CHANNEL)
+#define R__NERROR_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kError + LEVEL, CHANNEL)
+#define R__NWARNING_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kWarning + LEVEL, CHANNEL)
+#define R__NINFO_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kInfo + LEVEL, CHANNEL)
+#define R__NDEBUG_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kDebug + LEVEL, CHANNEL)
```
Comment (eguiraud) [inline core/foundation/inc/ROOT/RLogger.hxx:340]: maybe `OFFSET` rather than `LEVEL`?

**Review by eguiraud — CHANGES_REQUESTED** [review]
Nice! It's a "changes requested" just because there are a couple of wrong comments or leftover friend declarations/access specifiers to clean up, the rest is just minor comments.

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 6]
```diff
@@ -16,28 +16,136 @@
...
-   kInfo,    ///< Informational messages; used for instance for tracing
-   kWarning, ///< Warnings about likely unexpected behavior
-   kError,
-   kFatal
+   kOff = 0,
+   kFatal = 0x100,   ///< An error which causes further processing to be unreliable
+   kError = 0x200,   ///< An error
+   kWarning = 0x300, ///< Warnings about likely unexpected behavior
+   kInfo = 0x400,    ///< Informational messages; used for instance for tracing
+   kDebug = 0x500,   ///< Debug information; only useful for developers
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:37]: Yeah one can now decrease priority of `kError` by making it `kError + 100` or whatever, up to 255. I decided to order them this way around because of "verbosity": more verbose means higher cut-off means `kDebug` must be greater than `kError`. Let me know if you have an opinion on this, i.e. whether it's less surprising to deal with those as "priorities" / "severity" for messages, instead of "verbosity" for filtering. This enum will be used for both, so it's a question of what is more visible / relevant to users. Also @pcanal

File: core/foundation/src/RLogger.cxx [hunk 7]
```diff
@@ -46,14 +50,68 @@ bool RLogHandlerDefault::Emit(const ROOT::Experimental::RLogEntry &entry)
...
+      fChannel->SetVerbosity(fPrevLevel);
+   else
+      fLogManager->SetVerbosity(fPrevLevel);
+}
+
+RLogDiagCount::RLogDiagCount()
+{
+   std::atomic_init(&fNumWarnings, 0ll);
+   std::atomic_init(&fNumErrors, 0ll);
+   std::atomic_init(&fNumFatalErrors, 0ll);
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/src/RLogger.cxx:71]: Interesting - that seems to work indeed with all supported compilers. I wonder what made me use `atomic_init`?!

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 8]
```diff
@@ -16,28 +16,136 @@
...
+};
+
+
+/**
+ Keep track of emitted errors and warnings.
+ */
+class RLogDiagCount {
+   friend class RLogManager;
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:45]: No because of `++fChannel->fWarning`. I have switched to public methods instead.

File: core/foundation/inc/ROOT/RLogger.hxx (hunk 5 above)
Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:340]: `SEVERITY`? ("Offset" to what, and it just says it's a difference but doesn't come with a meaning / "unit")

Comment (eguiraud) [reply core/foundation/inc/ROOT/RLogger.hxx:340]: sure, it was just `LEVEL` that was overloaded

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 9]
```diff
@@ -16,28 +16,62 @@
 #ifndef ROOT7_RLogger
 #define ROOT7_RLogger
 
-#include <array>
+#include <atomic>
 #include <memory>
+#include <mutex>
 #include <sstream>
 #include <vector>
```
Comment (eguiraud) [inline core/foundation/inc/ROOT/RLogger.hxx:23]:
```suggestion
#include <vector>
#include <string>
```

to be very pedantic something for `std::swap` should also be included (and to be even more pedantic something for `std::move` :D )

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 10]
```diff
@@ -82,90 +127,236 @@ public:
...
+ */
+class RLogChannel : public RLogDiagCount {
+   /// Name as shown in diagnostics
+   std::string fName;
+
+   /// Verbosity of this channel. By default, use the global verbosity.
+   ELogLevel fVerbosity = ELogLevel::kUnset;
+
+public:
+   /// Construct a throw-away channel, with no channel name and no way of tweaking its verbosity it
```
Comment (eguiraud) [inline core/foundation/inc/ROOT/RLogger.hxx:145]:
```suggestion
   /// Construct a throw-away channel, with no channel name and no way of tweaking its verbosity
```

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 11]
```diff
@@ -82,90 +127,236 @@ public:
...
+   /// but with global suppression.
```
Comment (eguiraud) [inline core/foundation/inc/ROOT/RLogger.hxx:146]: > no way of tweaking its verbosity

can't I just call `SetVerbosity` on it?

File: core/foundation/src/RLogger.cxx [hunk 12]
```diff
@@ -14,46 +14,101 @@
...
-   if (!entry.fFile.empty())
-      strm << " " << entry.fFile << ':' << entry.fLine;
-   if (!entry.fFuncName.empty())
-      strm << " in " << entry.fFuncName;
+   if (!entry.fLocation.fFile.empty())
+      strm << " " << entry.fLocation.fFile << ':' << entry.fLocation.fLine;
+   if (!entry.fLocation.fFuncName.empty())
+      strm << " in " << entry.fLocation.fFuncName;
 
    static constexpr const int errorLevelOld[] = {0, 1000, 2000, 3000, 6000};
```
Comment (eguiraud) [inline core/foundation/src/RLogger.cxx:51]: I don't think these are correct but I might be wrong: we have a `kUnset`, don't have a `kPrint`, `kFatal` is element `1` in our enum but iiuc it's 6000 in `errorLevelOld`...

File: core/foundation/src/RLogger.cxx [hunk 13]
```diff
@@ -14,46 +14,101 @@
...
+         AddError();
+         if (channel)
+            channel->AddError();
+      } else if (entry.IsFatal()) {
+         AddFatalError();
+         if (channel)
+            channel->AddFatalError();
+      }
+   };
+   updateCounts();
```
Comment (eguiraud) [inline core/foundation/src/RLogger.cxx:90]: why do we need a lambda?

**Review by eguiraud — APPROVED** [review]
Sorry for being pedantic :smile: but I bet this stuff will get used _a lot_ for very long.

Looks good to me besides the few comments above. I guess any issue that might remain is minor and can be fixed later.

**Comment by Axel-Naumann** [comment]
No worries, thanks for doing that, IMO the code is actually improving by each of your reviews :-)

**Comment by Axel-Naumann** [comment]
Needs some more work as suggested by Philippe:
- use namespace-scope functions with local statics instead of global channel variables (as agreed on by @eguiraud but I forgot...);
- fwd declare that channel function and even its return type `RLogChannel` in the relevant headers, to reduce `#include` "load";

And #6931 which will go into a subsequent PR.

File: core/foundation/src/RLogger.cxx [hunk 14]
```diff
@@ -14,46 +14,109 @@
...
+
+bool RLogManager::Emit(const RLogEntry &entry)
+{
+   auto channel = entry.fChannel;
+
+   Increment(entry.fLevel);
+   if (channel)
+      channel->Increment(entry.fLevel);
+
+   // Is there a sepcific level for the channel? If so, take that,
```
Comment (eguiraud) [inline core/foundation/src/RLogger.cxx:100]:
```suggestion
   // Is there a specific level for the channel? If so, take that,
```

**Review by eguiraud — APPROVED** [review]
LGTM!

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 15]
```diff
@@ -55,118 +93,275 @@ public:
...
+   /// Construct the scoped count given a channel's counter.
+   /// The counter's lifetime must exceed the lifetime of this object!
+   RLogScopedDiagCount(RLogDiagCount &cnt)
+      : fCounter(&cnt), fInitialWarnings(cnt.GetNumWarnings()), fInitialErrors(cnt.GetNumErrors()),
+        fInitialFatalErrors(cnt.GetNumFatalErrors())
+   {
+   }
+
+   /// Construct the scoped count for any diagnostic, whatever its channel.
+   RLogScopedDiagCount() : RLogScopedDiagCount(RLogManager::Get()) {}
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:320]: Now sure about this one? I think it's fine to pass `{}` where a `RLogScopedDiagCount` is expected?

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 16]
```diff
@@ -55,118 +93,275 @@ public:
...
-     R__INFO_HERE("CodeGroupForInstanceLibrary") << "All we know is " << 42;
+ Builds a diagnostic entry, emitted by the static RLogManager upon destruction of this builder,
+ where - by definition - the RLogEntry has been completely built.
+
+ This builder can be used through the utility preprocessor macros R__ERROR_HERE,
+ R__WARNING_HERE etc like this:
+ ```
+     R__INFO_HERE("CodeChannelForInstanceLibrary") << "All we know is " << 42;
+     const int decreasedInfoLevel = 5;
+     R__NINFO_HERE("CodeChannelForInstanceLibrary", decreasedInfoLevel) << "All we know is " << 42;
```
Comment (eguiraud) [inline core/foundation/inc/ROOT/RLogger.hxx:221]: this is not a thing

File: core/foundation/src/RLogger.cxx [hunk 17]
```diff
@@ -14,46 +14,109 @@
...
-class RLogHandlerDefault: public ROOT::Experimental::RLogHandler {
+class RLogHandlerDefault : public RLogHandler {
 public:
    // Returns false if further emission of this log entry should be suppressed.
-   bool Emit(const ROOT::Experimental::RLogEntry &entry) override;
+   bool Emit(const RLogEntry &entry) override;
 };
 
-bool RLogHandlerDefault::Emit(const ROOT::Experimental::RLogEntry &entry)
+bool RLogHandlerDefault::Emit(const RLogEntry &entry)
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/foundation/src/RLogger.cxx:37]: Because the class definition would span a 20+ line function - and that's beyond my personal limit for in-class definitions...

**Comment by Axel-Naumann** [comment]
I fixed the misconfig of the Mac.

File: core/foundation/inc/ROOT/RLogger.hxx [hunk 18]
```diff
@@ -176,13 +326,43 @@ public:
...
-#define R__INFO_HERE(GROUP) R__LOG_HERE(ROOT::Experimental::ELogLevel::kInfo, GROUP)
-#define R__DEBUG_HERE(GROUP) R__LOG_HERE(ROOT::Experimental::ELogLevel::kDebug, GROUP)
+/*
+ Some implementation details:
+
+ - The conditional `RLogBuilder` use prevents stream operators from being called if
+ verbosity is too low, i.e.:
+ ````
+ RLogScopedVerbosity silence(RLogLevel::kFatal);
+ R__LOG_DEBUG() << WillNotBeCalled();
```
Comment (pcanal) [inline core/foundation/inc/ROOT/RLogger.hxx:336]: This comment/doc might not be matching the code anymore.

Comment (Axel-Naumann) [reply core/foundation/inc/ROOT/RLogger.hxx:336]: Thanks, fixed in 5528d31fbf1a004a580dbe56d85e7a6d077744d7
